File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    /*
    556698111222   count =2
                  
   max(j-i)
   max(j-i+1)
    */
    int totalFruit(vector<int>& fruits) {
     unordered_map<int,int>mp;
        int n=fruits.size();
        int i=0,j, m=0;
        for(j=0; j<n; j++){
            mp[fruits[j]]++; //adding to map
            while(mp.size()>2){// check if fruits > 2               
                if(--mp[fruits[i]]==0) mp.erase(fruits[i]); //erase both the counts and the elements
                i++;
            }   
             m=max(m,j-i+1);
        }
        return m;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        maxi = 0
        fruits_dict = defaultdict(int)
        start = 0
        length = len(fruits)
        for end in range(length):
            fruits_dict[fruits[end]] += 1           
            while len(fruits_dict) > 2:
                fruits_dict[fruits[start]] -= 1
                if fruits_dict[fruits[start]]  == 0:
                    fruits_dict.pop(fruits[start])
                start += 1
            maxi = max(maxi, end - start + 1)  
            
        return maxi
                